File: Easy/94_binary_tree_inorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    Wrapper wrapper = new Wrapper();
    public List<Integer> inorderTraversal(TreeNode root) {  
        traversal(root);
        return wrapper.list;
    }
    
    public void traversal(TreeNode root) {
        if (root == null) return;
        traversal(root.left);
        wrapper.list.add(root.val);
        traversal(root.right);
        
    }
}

class Wrapper {
    public ArrayList<Integer> list;
    public Wrapper() {
        list = new ArrayList<>();
    }
}

/**
* Runtime: 0 ms, faster than 100.00% of Java online submissions for Binary Tree Inorder Traversal.
* Memory Usage: 38.7 MB, less than 32.25% of Java online submissions for Binary Tree Inorder Traversal.
*/

File: Rust/Easy/94_binary_tree_inorder_traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut answer: Vec<i32> = vec![];
        Self::dfs(&root, &mut answer);
        answer
    }

    pub fn dfs(root: &Option<Rc<RefCell<TreeNode>>>, answer: &mut Vec<i32>) -> () {
        match root {
            Some(r) => {
                let mut left = Self::dfs(&r.borrow().left, answer);
                answer.push(r.borrow().val);
                let mut right = Self::dfs(&r.borrow().right, answer);
            },
            None => return,
        }
    }
}
